tiledb/ml/data_apis/tensorflow_sparse.py

```diff
@@ -0,0 +1,154 @@
+        :param x_array: TileDB Sparse Array. Array that contains features.
+        :param y_array: TileDB Dense/Sparse Array. Array that contains labels.
+        :param batch_size: Integer. The size of the batch that the implemented _generator method will return.
+        """
+
+        cls.x = x_array
+        cls.y = y_array
+
+        # Check that x and y have the same number of rows
+        if x_array.schema.domain.shape[0] != y_array.schema.domain.shape[0]:
+            raise Exception(
+                "X and Y should have the same number of rows, i.e., the 1st dimension "
+                "of TileDB arrays X, Y should be of equal domain extent."
+            )
+
+        # Get number of observations
+        rows = x_array.schema.domain.shape[0]
+
+        # Get x and y shapes
+        cls.x_shape = (None,) + x_array.schema.domain.shape[1:]
+        cls.y_shape = (None,) + y_array.schema.domain.shape[1:]
+
+        # Get x and y data types
+        cls.x_dtype = x_array.schema.attr(0).dtype
+        cls.y_dtype = y_array.schema.attr(0).dtype
+
+        if isinstance(y_array, tiledb.SparseArray):
+            return dataset_ops.Dataset.from_generator(
+                generator=cls._generator_sparse_sparse,
+                output_signature=(
+                    tf.SparseTensorSpec(shape=cls.x_shape, dtype=cls.x_dtype),
+                    tf.SparseTensorSpec(shape=cls.y_shape, dtype=cls.y_dtype),
+                ),
+                args=(rows, batch_size),
+            )
+        else:
+            return dataset_ops.Dataset.from_generator(
+                generator=cls._generator_sparse_dense,
+                output_signature=(
+                    tf.SparseTensorSpec(shape=cls.x_shape, dtype=cls.x_dtype),
+                    tf.TensorSpec(shape=cls.y_shape, dtype=cls.y_dtype),
+                ),
+                args=(rows, batch_size),
+            )
+
+    @classmethod
+    def _generator_sparse_sparse(cls, rows: int, batch_size: int) -> tuple:
+        """
+        A generator function that yields the next training batch.
+        :param x: TileDB array. An opened TileDB array which contains features.
+        :param y: TileDB array. An opened TileDB array which contains labels.
+        :param rows: Integer. The number of observations in x, y datasets.
+        :param batch_size: Integer. Size of batch, i.e., number of rows returned per call.
+        :return: Tuple. Tuple that contains x and y batches.
+        """
+        # Loop over batches
+        # https://github.com/tensorflow/tensorflow/issues/44565
+        for offset in range(0, rows, batch_size):
+            y_batch = cls.y[offset : offset + batch_size]
+            x_batch = cls.x[offset : offset + batch_size]
+
+            # TODO: Both for dense case support multiple attributes
+            values_y = list(y_batch.items())[0][1]
+            values_x = list(x_batch.items())[0][1]
```

Comment:
Unfortunately `list(y_batch.values())[0]` returns the `name` of the attribute in our case the `features` since the `y_batch.values()` returns an `OrderedDict` object. Getting `[0][1]` takes the values of the `feature` and the `TODO` comment is there to remind us that this should be generalised in case of multiple feature data.

Reply:
Ah my bad, thought `y_batch` was the `OrderedDict`